src/main/java/digital/slovensko/autogram/core/visualization/DocumentVisualizationBuilder.java

```diff
@@ -77,6 +94,79 @@ private Visualization createVisualization(SigningJob job, MimeType transformatio
         return new UnsupportedVisualization(job);
     }
 
+    private DSSDocument getOriginalDocument(DSSDocument document) {
+        SignedDocumentValidator documentValidator = SignedDocumentValidator.fromDocument(document);
+        documentValidator.setCertificateVerifier(new CommonCertificateVerifier());
+        List<AdvancedSignature> signatures = documentValidator.getSignatures();
+        if (signatures.isEmpty()) {
+            throw new RuntimeException("No signatures in document");
+        }
+        AdvancedSignature advancedSignature = signatures.get(0);
+        List<DSSDocument> originalDocuments = documentValidator.getOriginalDocuments(advancedSignature.getId());
+        if (originalDocuments.isEmpty()) {
+            throw new RuntimeException("No original documents found");
+        }
+        return originalDocuments.get(0);
+    }
+
+    private void setMimeTypeFromManifest(DSSDocument originalDocument, DSSDocument document) {
+        DSSDocument manifest = getManifest(originalDocument);
+        if (manifest == null) {
+            return;
+        }
+
+        String documentName = document.getName();
+        MimeType mimeType = getMimeTypeFromManifest(manifest, documentName);
+        if (mimeType == null) {
+            return;
+        }
+
+        document.setMimeType(mimeType);
+    }
+
+    private DSSDocument getManifest(DSSDocument originalDocument) {
+        ASiCWithXAdESContainerExtractor extractor = new ASiCWithXAdESContainerExtractor(originalDocument);
+        ASiCContent aSiCContent = extractor.extract();
+        List<DSSDocument> manifestDocuments = aSiCContent.getManifestDocuments();
+        if (manifestDocuments.isEmpty()) {
+            return null;
+        }
+        return manifestDocuments.get(0);
+    }
+
+    private MimeType getMimeTypeFromManifest(DSSDocument manifest, String documentName) {
+        NodeList fileEntries = getFileEntriesFromManifest(manifest);
+        if (fileEntries == null) {
+            return null;
+        }
+
+        for (int i = 0; i < fileEntries.getLength(); i++) {
```

Comment:
keď nepoužívame ten int i, tak by som dal `for (var item : fileEntries)`

Reply:
```fileEntries``` neni Collection, ale NodeList a nad nim mozes zavolat len ```item(i)``` a ```getLength()```